### PR TITLE
Hide gridlines and ticks overlapping inside ticklabels

### PR DESCRIPTION
**Pull request by archmoj**
Addressing #5502.

@plotly/plotly_js

### CODE REVIEW

File: src/plots/cartesian/axes.js [hunk 1]
```diff
@@ -2861,7 +2867,9 @@ axes.drawGrid = function(gd, ax, opts) {
     grid.attr('transform', opts.transFn)
         .attr('d', opts.path)
         .call(Color.stroke, ax.gridcolor || '#ddd')
-        .style('stroke-width', ax._gw + 'px');
+        .style('stroke-width', ax._gw + 'px')
+        .style({ opacity: 100 }); // ensure visible
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:2871]:
```suggestion
        .style('opacity', 1); // ensure visible
```
- use the 2-argument form, as that's the only form available in later d3 versions (same comment elsewhere in this PR)
- opaque is 1, not 100, right?

**Comment by alexcjohnson** [comment]
As described in #5502 we should also be removing the ticks & tick labels at these positions, not just the gridlines. Perhaps that argues for a simpler solution where the range over which ticks are generated is reduced?

File: src/plots/cartesian/axes.js (hunk 1 above)
Comment (archmoj) [reply src/plots/cartesian/axes.js]: Addressed in https://github.com/plotly/plotly.js/pull/5550/commits/60fe3d2a39bffffc70596b757b42048d635b02d8.

**Comment by nicolaskruchten** [comment]
Does this PR include the "addendum" scope from https://github.com/plotly/plotly.js/issues/5502#issuecomment-800437485 ? After speaking with the customer, it's OK for X ticklabels to slightly underhang Y inside-ticklabels, so we convened that the centerline was a better trip-point than either edge.

**Comment by alexcjohnson** [comment]
I see https://github.com/plotly/plotly.js/pull/5550/commits/769c7b899b360f70853b213f70357769868972cf that indicates at one point we were removing zero lines that intersect these labels, but now we're not anymore... don't we want to? My understanding was that the goal here is that none of the counteraxis marks at a location beyond the beginning of these inside labels should be visible.

Also I see some labels (`0` on the y axis in this example from `axes_chain_scaleanchor_matches2_inside-ticklabels.png`) appear on the other side of a removed label (`0.5`) - this seems distinct from the zero line question, as other `0` labels in the same plot are removed.
Screen Shot 2021-03-19 at 7 06 03 PM

**Comment by nicolaskruchten** [comment]
We do want to remove the zero lines that intersect, yes

**Comment by nicolaskruchten** [comment]
Can I get a Pen for this please?

**Comment by alexcjohnson** [comment]
Two bits of behavior here look not quite right to me:
- it's possible for a set of marks to disappear when the value goes behind the labels, but then to come back between the label and the edge of the plot. This can happen with no ticks or outside ticks, just due to the little gap between labels and the edge, but it's more prominent if you put long ticks inside.

Here the labels for -8 and 9 are hidden, as they should be:
hidden labels
But then they come back (note the labels rarely come back when they're inside, unless you have super long inside ticks, but the ticks and gridlines come back):
visible again
Seems to me the correct logic should be "past the beginning of the label bounding box" rather than "inside the label bounding box"

- While I'm dragging, the gridlines and zerolines disappear at the right time but the ticks and tick labels behave incorrectly. Ticks that were hidden when I started dragging don't appear until I mouse up, whereas all tick labels (on both axes, the dragged and the static) appear when I start dragging, then on mouseup some are removed.

Here (while dragging) you see a label but no tick because it was hidden pre-drag:
no tick on 8
And here (still while dragging) you see both 8 and 9 labels on top of each other:
8 and 9
But then they disappear on mouse up:
correct on mouseup

File: src/plots/cartesian/axes.js [hunk 2]
```diff
@@ -34,6 +34,14 @@ var ONESEC = constants.ONESEC;
 var MINUS_SIGN = constants.MINUS_SIGN;
 var BADNUM = constants.BADNUM;
 
+var VISIBLE = { display: null };
+var HIDDEN = { display: 'none' };
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:38]: Per https://github.com/plotly/plotly.js/pull/5550#discussion_r594815159 we should not add new instances of the single-arg form of `selection.style`, these should all be `.style('display', null)` etc.

Comment (archmoj) [reply src/plots/cartesian/axes.js]: Addressed in https://github.com/plotly/plotly.js/pull/5550/commits/f05cd39ca199b0395ef340998cf5e977a1676f06.

**Comment by archmoj** [comment]
> * While I'm dragging, the gridlines and zerolines disappear at the right time but the ticks and tick labels behave incorrectly. Ticks that were hidden when I started dragging don't appear until I mouse up, whereas all tick labels (on both axes, the dragged and the static) appear when I start dragging, then on mouseup some are removed.

Good catch!
The second bit is addressed in https://github.com/plotly/plotly.js/pull/5550/commits/caf32396ba9d416abde3c27f2f8b19f72c1c7a0f.

**Comment by alexcjohnson** [comment]
> The second bit is addressed in caf3239.

That seems to fix the ticks, but I'm still seeing some labels that should have been removed.
Screen Shot 2021-03-31 at 12 29 09 PM (2)

**Comment by archmoj** [comment]
> > The second bit is addressed in [caf3239](https://github.com/plotly/plotly.js/commit/caf32396ba9d416abde3c27f2f8b19f72c1c7a0f).
> 
> That seems to fix the ticks, but I'm still seeing some labels that should have been removed.
> Screen Shot 2021-03-31 at 12 29 09 PM (2)

I can't replicate this problem.

**Comment by alexcjohnson** [comment]
Here's panning on one of the subplots of `axes_chain_scaleanchor_matches_inside-ticklabels` - the labels in the lower left corner appear (overlapping each other) as soon as I start to drag, and disappear only on mouseup
labels hide on mouseup

**Comment by archmoj** [comment]
> Here's panning on one of the subplots of `axes_chain_scaleanchor_matches_inside-ticklabels` - the labels in the lower left corner appear (overlapping each other) as soon as I start to drag, and disappear only on mouseup
> labels hide on mouseup

I am not 100% sure if we could easily fix that at the moment considering complexity of displaying and hiding of inside ticklabels. 
But we could certainly try it on a separate PR, IMHO.

**Comment by alexcjohnson** [comment]
@archmoj give it a try in this PR - it's an edge case but we really don't want to be introducing a known bug in one PR just to remove it in another.

**Comment by archmoj** [comment]
> @archmoj give it a try in this PR - it's an edge case but we really don't want to be introducing a known bug in one PR just to remove it in another.

Good call. Addressed in https://github.com/plotly/plotly.js/pull/5550/commits/e683b07ff5dff0370d945b9180376ebf10e0e312.

**Review by alexcjohnson — APPROVED** [review]
Love it 💃

File: src/plots/cartesian/axes.js [hunk 3]
```diff
@@ -3025,8 +3043,7 @@ axes.drawLabels = function(gd, ax, opts) {
                 });
 
                 if(isInside) {
-                    // ensure visible
-                    thisText.style({ opacity: 100 });
+                    thisText.style('opacity', 0); // visible
```
Comment (archmoj) [inline src/plots/cartesian/axes.js:3046]: Opacity should be set to 1 here not 0.
I'll open a PR to fix this.

Comment (archmoj) [reply src/plots/cartesian/axes.js:3046]: Fixed by #5583.

**Pull request merged**